src/sentry/api/event_search.py

```diff
@@ -1365,233 +1365,354 @@ def get_default(self, params):
         return int(interval)
 
 
+class Function(object):
+    def __init__(
+        self,
+        name,
+        required_args=None,
+        optional_args=None,
+        calculated_args=None,
+        column=None,
+        aggregate=None,
+        transform=None,
+        result_type=None,
+    ):
+        """
+        Specifies a function interface that must be followed when defining new functions
+
+        :param str name: The name of the function, this refers to the name to invoke.
+        :param list[FunctionArg] required_args: The list of required arguments to the function.
+            If any of these arguments are not specified, an error will be raised.
+        :param list[FunctionArg] optional_args: The list of optional arguments to the function.
+            If any of these arguments are not specified, they will be filled using their default value.
+        :param list[obj] calculated_args: The list of calculated arguments to the function.
+            These arguments will be computed based on the list of specified arguments.
+        :param [str, [any], str or None] column: The column to be passed to snuba once formatted.
+            The arguments will be filled into the column where needed. This must not be an aggregate.
+        :param [str, [any], str or None] aggregate: The aggregate to be passed to snuba once formatted.
+            The arguments will be filled into the aggregate where needed. This must be an aggregate.
+        :param str transform: NOTE: Use aggregate over transform whenever possible.
+            An aggregate string to be passed to snuba once formatted. The arguments
+            will be filled into the string using `.format(...)`.
+        :param str result_type: The resulting type of this function. Can be any of the following
+            (duration, string, number, integer, percentage, date).
+        """
```

Comment:
This is much simpler to read than the previous implementation. Great work :clap: